tsl/src/continuous_aggs/create.c

```diff
@@ -1173,39 +1183,148 @@ cagg_validate_query(const Query *query, const bool finalized, const char *cagg_s
+						ereport(ERROR,
+								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								 errmsg("invalid continuous aggregate view"),
+								 errdetail(
+									 "joins with using clause in continuous aggregate definition"
+									 " work for postgres versions 13 and above")));
+#endif
+					jointype = join->jointype;
+					op = (OpExpr *) join->quals;
+					rte = list_nth(query->rtable, ((RangeTblRef *) join->larg)->rtindex - 1);
+					rte2 = list_nth(query->rtable, ((RangeTblRef *) join->rarg)->rtindex - 1);
+				}
+			}
+		}
+		if ((rte->relkind != RELKIND_RELATION && rte->relkind != RELKIND_VIEW) ||
+			rte->tablesample || rte->inh == false)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("invalid continuous aggregate view"),
+					 errdetail("from clause can only have one hypertable and one normal table")));
+
+		if ((rte2->relkind != RELKIND_RELATION && rte2->relkind != RELKIND_VIEW) ||
+			rte2->tablesample || rte2->inh == false)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("invalid continuous aggregate view"),
+					 errdetail("from clause can only have one hypertable and one normal table")));
+		/*
+		 * Cagg with joins does not allow nested caggs in from clause.
+		 */
+		if (rte->relkind == RELKIND_VIEW || rte2->relkind == RELKIND_VIEW)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("joins for nested continuous aggregates are not supported")));
+		/* Both tables could not be hypertables or normal table */
+		if (ts_is_hypertable(rte->relid) == ts_is_hypertable(rte2->relid))
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("only one hypertable and one normal table are allowed"
+							" in continuous aggregates")));
+		/* Only inner joins are allowed */
+		if (jointype != JOIN_INNER)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("only inner joins are supported in continuous aggregates")));
+		/* Only equality conditions are permitted on joins */
+		if (IsA(op, OpExpr) && list_length(castNode(OpExpr, op)->args) == 2)
+		{
+			Oid left_type = exprType(linitial(op->args));
+			Oid right_type = exprType(lsecond(op->args));
+			if (!ts_is_equality_operator(op->opno, left_type, right_type))
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("invalid continuous aggregate view"),
+						 errdetail(
+							 "only equality conditions are supported in continuous aggregates")));
+		}
+		else
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("invalid continuous aggregate view")));
+
+		/*
+		 * Record the table oid of the normal table, this is required so
+		 * that we know which one is hypertable to carry out the related
+		 * processing in later parts of code.
+		 */
+		normal_table_id = ts_is_hypertable(rte->relid) ? rte2->relid : rte->relid;
+		if (normal_table_id == rte->relid)
+			rte = rte2;
```

Comment:
I'm a bit confused with this logic... maybe we should clarify a bit more the comments.

Reply:
In the later part of the code (from line 1326 onwards) we are using rte to get the cagg information, etc. for dimension info and time_bucket info  etc. and for that this needs to be the rte of the hypertable. Since I did not want to touch that part of code unnecessarily, hence kept the name rte for this and that's why switching the rte in case it corresponds to the normal table

Reply:
In this case a better variable names will help understand the code